### ISSUE TITLE
Unexpected output with CLFFT_REAL

### ISSUE THREAD

**Issue by PJMKT** [clMathLibraries/clFFT#78]
Hello,

I am having trouble with CLFFT_REAL transforms, depending on the size of the array. I tried with several versions of openCL (cuda-6.5 or cuda-7.0) and with the master and develop branches of clFFT.
I am using Ubuntu 14.04 and g++-4.8.2
I wrote a test program (see below) to highlight what's going wrong and here is the output:
```
CLFFT with type float and with layouts Real to Complex interleaved succeeded 
CLFFT with type double and with layouts Real to Complex interleaved succeeded 
CLFFT with type float and with layouts Real to Hermitian interleaved succeeded 
CLFFT with type double and with layouts Real to Hermitian interleaved failed.
```

By the way, changing the size of the array will not give exactly the same results - this is for N = 2.

Thank you.

---main.cpp:
```c++
#include <stdlib.h>
#include <iostream>
#include <clFFT.h>
#include <cmath>

const char * layoutToString( clfftLayout_ layout )
{
   switch( layout ) {
   case CLFFT_COMPLEX_INTERLEAVED: return "Complex interleaved";
   case CLFFT_COMPLEX_PLANAR: return "Complex planar";
   case CLFFT_HERMITIAN_INTERLEAVED: return "Hermitian interleaved";
   case CLFFT_HERMITIAN_PLANAR: return "Hermitian planar";
   case CLFFT_REAL: return "Real";
   case ENDLAYOUT: return "Unknown";
   }
   return "Not found";
}

template< class T >
void testCLFFT( clfftLayout_ startLayout, clfftLayout_ intermediaryLayout )
{
   bool success = true;
   const unsigned int N = 2;
   cl_int err;
   cl_platform_id platform = 0;
   cl_device_id device = 0;
   cl_context_properties props[3] = { CL_CONTEXT_PLATFORM, 0, 0 };
   cl_context _ctx = 0;
   cl_command_queue _queue = 0;
   cl_mem _bufferIn;
   T *X;

   clfftPlanHandle _planHandle;
   clfftDim clfftDim = CLFFT_1D;
   size_t clLengths[1] = {N};

   err = clGetPlatformIDs( 1, &platform, NULL );
   err = clGetDeviceIDs( platform, CL_DEVICE_TYPE_GPU, 1, &device, NULL );
   props[1] = (cl_context_properties)platform;
   _ctx = clCreateContext( props, 1, &device, NULL, NULL, &err );
   _queue = clCreateCommandQueue( _ctx, device, 0, &err );

   clfftSetupData fftSetup;
   err = clfftInitSetupData(&fftSetup);
   err = clfftSetup(&fftSetup);

// Filling array
   //    unsigned int N1 = N/2 + 1;
   X = (T *) malloc( N * 2 * sizeof(T) );
   for( unsigned int i = 0; i < 2 * N; ++i ) X[ i ] = static_cast< T >( 0 );
   for( unsigned int i = 0; i < N; ++i ) X[ i ] = static_cast< T >( i );

// Forward transform
   err = clfftCreateDefaultPlan(&_planHandle, _ctx, clfftDim, clLengths);
   if( sizeof(T) == 4 ) err = clfftSetPlanPrecision(_planHandle, CLFFT_SINGLE);
   else err = clfftSetPlanPrecision(_planHandle, CLFFT_DOUBLE);
   err = clfftSetLayout(_planHandle, startLayout, intermediaryLayout );
   err = clfftSetResultLocation(_planHandle, CLFFT_INPLACE);
   err = clfftBakePlan(_planHandle, 1, &_queue, NULL, NULL);
   _bufferIn = clCreateBuffer( _ctx, CL_MEM_READ_WRITE, N * 2 * sizeof(T), NULL, &err );
   err = clEnqueueWriteBuffer( _queue, _bufferIn, CL_TRUE, 0, N * 2 * sizeof(T), X, 0, NULL, NULL );
   err = clfftEnqueueTransform(_planHandle, CLFFT_FORWARD, 1, &_queue, 0, NULL, NULL, &_bufferIn, NULL, NULL);
   err = clFinish(_queue);
   err = clEnqueueReadBuffer( _queue, _bufferIn, CL_TRUE, 0, N * 2 * sizeof(T), X, 0, NULL, NULL );

// Backward transform
   err = clfftCreateDefaultPlan(&_planHandle, _ctx, clfftDim, clLengths);
   if( sizeof(T) == 4 ) err = clfftSetPlanPrecision(_planHandle, CLFFT_SINGLE);
   else err = clfftSetPlanPrecision(_planHandle, CLFFT_DOUBLE);
   err = clfftSetLayout(_planHandle, intermediaryLayout, startLayout );
   err = clfftSetResultLocation(_planHandle, CLFFT_INPLACE);
   err = clfftBakePlan(_planHandle, 1, &_queue, NULL, NULL);
   _bufferIn = clCreateBuffer( _ctx, CL_MEM_READ_WRITE, N * 2 * sizeof(T), NULL, &err );
   err = clEnqueueWriteBuffer( _queue, _bufferIn, CL_TRUE, 0, N * 2 * sizeof(T), X, 0, NULL, NULL );
   err = clfftEnqueueTransform(_planHandle, CLFFT_BACKWARD, 1, &_queue, 0, NULL, NULL, &_bufferIn, NULL, NULL);
   err = clFinish(_queue);
   err = clEnqueueReadBuffer( _queue, _bufferIn, CL_TRUE, 0, N * 2 * sizeof(T), X, 0, NULL, NULL );

// Checking we find the initial array values
   for(unsigned int i = 0; i < N; ++i) {
      if( X[i] != i ) {
         success =  false;
         break;
      }
   }

// Releasing memory
   clReleaseMemObject( _bufferIn );
   free(X);
   err = clfftDestroyPlan( &_planHandle );
   clfftTeardown( );
   clReleaseCommandQueue( _queue );
   clReleaseContext( _ctx );

// Displaying FFT success status
   std::cout << "CLFFT with type " << ( sizeof(T) == 4 ? "float" : "double") << " and with layouts " <<
         layoutToString( startLayout ) << " to " << layoutToString( intermediaryLayout ) <<
         ( success ? " succeeded " : " failed " ) << std::endl;
}

int main( void )
{
   testCLFFT< float >( CLFFT_REAL, CLFFT_COMPLEX_INTERLEAVED );
   testCLFFT< double >( CLFFT_REAL, CLFFT_COMPLEX_INTERLEAVED );
   testCLFFT< float >( CLFFT_REAL, CLFFT_HERMITIAN_INTERLEAVED );
   testCLFFT< double >( CLFFT_REAL, CLFFT_HERMITIAN_INTERLEAVED );
   return 0;
}
```

Compile line:
```
g++ main.cpp -m64 -pipe -O2 -Wno-unused-but-set-parameter -Wall -std=c++11 -I/usr/local/mkt-dev/install/cuda-7.0/include  -I/usr/local/include  -L/usr/local/lib64 -lclFFT -L/usr/local/mkt-dev/install/cuda-7.0/lib64 -lOpenCL  -o fft1D
```

**Comment by bragadeesh** [comment]
What OpenCL platform was this tried on? Also, the way you are testing for success/failure is incorrect. With floating point, you need at allow for small tolerances, the answers will not be bit exact generally. Can you print the results out and visually see if they are only off by a small margin or if the answers are   completely wrong.

**Comment by PJMKT** [comment]
Thank you for your answer.

Sorry for the late answer, I realized that clGetPlatformInfo returned CL_INVALID_PLATFORM, so I asked a colleague of mine to try on his machine. I am waiting for his tests.

To respond to your second question, I initially tested with a little margin allowed; but for code clarity, and because for small N, it does seem to matter, I simplified the test. The output array is 0 and 0 instead of 0 and 1, in the last case.
The problem seems to generally occur at N/2 + 1. For N = 4 :
```
0: 0
1: 1
2: 1.59098e-314
3: 3
```
I'll let you know when I know more, but it might be a platform problem as you suggested.
Cheers.
